src/riak_kv_ddl.erl

```diff
@@ -0,0 +1,612 @@
+%% -------------------------------------------------------------------
+%%
+%% riak_kv_ddl: API module for the DDL
+%%
+%%
+%% Copyright (c) 2007-2015 Basho Technologies, Inc.  All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+-module(riak_kv_ddl).
+
+-include("riak_kv_index.hrl").
+-include("riak_kv_ddl.hrl").
+
+%% this function can be used to work out which Module to use
+-export([
+	 make_module_name/1
+	]).
+
+%% a helper function for destructuring data objects
+%% and testing the validity of field names
+%% the generated helper functions cannot contain
+%% record definitions because of the build cycle
+%% so this function can be called out to to pick
+%% apart the DDL records
+-export([
+	 get_partition_key/2,
+	 get_local_key/2,
+	 is_valid_field/2,
+	 is_query_valid/2
+	]).
+
+-ifdef(TEST).
+%% for debugging only
+-export([
+	 make_ddl/2,
+	 are_selections_valid/3
+
+	]).
+-endif.
+
```

Comment:
These functions might be useful to have available but it might be awkward to recompile. Might be worth putting them into a separate module; ddl_debug or something similar and not have them macro'ed out.